### PR TITLE
fix(core): fix disk space leak on upserting with dedup

### PR DESCRIPTION
**Pull request by ideoma**
Fixes https://github.com/questdb/questdb/issues/4917

When the same data is inserted into a table with deduplication, the table size should not grow. This is fixed in this PR.

Although the changes look logical and trivial they are in a highly vulnerable area hence the change should be set for a prolonged period of fuzz testing before making it into a release.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 30 (96.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|21|22|95.45%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/LongSequenceFunctionFactory.java|7|7|100.00%|

File: core/src/main/java/io/questdb/std/Files.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public final class Files {
     private static final AtomicInteger fdCounter = new AtomicInteger();
     private static final LongHashSet openFds = new LongHashSet();
     // To be set in tests to check every call for using OPEN file descriptor
-    public static boolean PARANOIA_FD_MODE = false;
+    public static boolean PARANOIA_FD_MODE = true;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/Files.java:68]: debris?

File: core/src/main/java/io/questdb/cairo/O3OpenColumnJob.java [hunk 2]
```diff
@@ -460,21 +446,45 @@ public static void o3PartitionMerge(
...
+            dstDataSize = dstDataAppendOffset2 + suffixSize + suffixSize2;
+
+            dstDataFd = openRW(ff, dFile(pathToNewPartition.trimTo(pplen), columnName, columnNameTxn), LOG, tableWriter.getConfiguration().getWriterFileOpenOpts());
+            if (dstDataSize > 0) {
+                dstVarAddr = mapRW(ff, dstDataFd, dstDataSize, MemoryTag.MMAP_O3);
+                if (!mixedIOFlag) {
+                    ff.madvise(dstVarAddr, dstDataSize, Files.POSIX_MADV_RANDOM);
+                }
+            } else {
+                dstVarAddr = 0;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3OpenColumnJob.java:485]: static analysis warning: redundant assignment

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/O3OpenColumnJob.java:485]: there is also an unused variable warning above

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 29 (96.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|20|21|95.24%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/LongSequenceFunctionFactory.java|7|7|100.00%|

**Pull request merged**